Go: Deal with incorrect toolchain versions

mbg opened the pull request:
Starting in Go 1.21, Go uses a new version format for toolchains which uses [the `1.N.P` syntax](https://tip.golang.org/doc/toolchain#version). This is a problem for our users when their `go.mod` files only contain a `go` directive that uses the old version format, e.g. `go 1.22`. `1.22` is a valid _language_ version, but not a valid _toolchain_ version. If there is no additional `toolchain` directive in the `go.mod` file, then `go` uses the _language_ version as the _toolchain_ version. In the example, `1.22` is _not_ a valid _toolchain_ version and certain `go` commands, particularly those which try to acquire the toolchain corresponding to that version, fail with an unhelpful error.

In this PR, we make two changes to the Go autobuilder:

1. We detect `go.mod` files where the _language_ version is `1.21` or greater, doesn't use the new toolchain version syntax, and there is no explicit `toolchain` directive. We emit a warning-level diagnostic in this case.
2. We explicitly invoke `go version` with `GOTOOLCHAIN=go1.N.P` (i.e. the correct representation of the toolchain version) to get `go` to install that version and avoid subsequent errors despite the incorrect version format. As an optimisation for repos with a lot of `go.mod` files, we now keep track of toolchain versions that we know are installed and only do this if we don't yet know that the corresponding version is installed.

smowton reviewed (commented):
If you want to test an actual download, consider setting `GOPATH` to an empty directory so it certainly won't find an existing toolchain downloaded by a prior integration test or similar.

mbg commented:
@smowton We already set a custom `GOPATH` for every integration test (as of https://github.com/github/codeql/pull/15355) and that doesn't seem to affect which toolchains `go` can/cannot find. I also experimented with overriding `GOTOOLCHAIN` to force a particular toolchain, but couldn't get that to work right because we eventually want to switch to the newer one during the test and there seems to be no good way to get that to switch at the right point.

owen-mc commented on go/extractor/diagnostics/diagnostics.go:
```diff
@@ -508,3 +508,18 @@ func EmitExtractionFailedForProjects(path []string) {
+func EmitInvalidToolchainVersion(goModPath string, version string) {
+	emitDiagnostic(
+		"go/autobuilder/invalid-go-toolchain-version",
+		fmt.Sprintf("`%s` is not a valid Go toolchain version", version),
```
It seems this should part should always stay the same. I suggest `Invalid Go toolchain version`. You may as well fix `EmitExtractionFailedForProjects` above at the same time, to something like `Unable to extract some Go projects`.

mbg commented:
@smowton weirdness seems to have been due to the awful `semver` behaviour that requires versions to start with a `v` character so the changes in `toolchain.go` where we kept track of installed Go versions always ended up inserting/checking for `""` in the set since `semver.Canonical` returns `""` if the version is not valid.

I have fixed those bugs and added an integration test.

owen-mc commented on go/extractor/toolchain/toolchain_test.go:
```diff
@@ -14,3 +14,9 @@ func TestParseGoVersion(t *testing.T) {
+
+func TestHasGoVersion(t *testing.T) {
+	if HasGoVersion("1.21") {
+		t.Error("Exepected HasGoVersion(\"1.21\" to be false, but got true)")
```
```suggestion
		t.Error("Exepected HasGoVersion(\"1.21\") to be false, but got true")
```

owen-mc commented on go/extractor/toolchain/toolchain_test.go:
```diff
@@ -14,3 +14,9 @@ func TestParseGoVersion(t *testing.T) {
+
+func TestHasGoVersion(t *testing.T) {
+	if HasGoVersion("1.21") {
+		t.Error("Exepected HasGoVersion(\"1.21\") to be false, but got true")
```
```suggestion
		t.Error("Expected HasGoVersion(\"1.21\") to be false, but got true")
```

owen-mc replied:
I'm sorry, I can't ignore it, even though it's just a test.
